src/enzo/Grid_ActiveParticleHandler.C

```diff
@@ -0,0 +1,249 @@
+int GetUnits(float *DensityUnits, float *LengthUnits,
+	     float *TemperatureUnits, float *TimeUnits,
+	     float *VelocityUnits, FLOAT Time);
+int FindField(int field, int farray[], int numfields);
+ 
+int grid::ActiveParticleHandler(HierarchyEntry* SubgridPointer, int level,
+                                float dtLevelAbove, int &NumberOfNewParticles)
+{
+
+  if (EnabledActiveParticlesCount == 0) return SUCCESS;
+
+  if (MyProcessorNumber != ProcessorNumber)
+    return SUCCESS;
+ 
+  if (NumberOfBaryonFields == 0)
+    return SUCCESS;
+
+  /* First, set under_subgrid field */
+  HierarchyEntry *Subgrid;
+  this->ZeroSolutionUnderSubgrid(NULL, ZERO_UNDER_SUBGRID_FIELD);
+  for (Subgrid = SubgridPointer; Subgrid; Subgrid = Subgrid->NextGridThisLevel)
+    this->ZeroSolutionUnderSubgrid(Subgrid->GridData, ZERO_UNDER_SUBGRID_FIELD);
+
+  /* initialize */
+
+  LCAPERF_START("grid_ActiveParticleHandler");
+
+  /* First we identify the data dependencies */
+
+  struct ActiveParticleFormationDataFlags flags = flags_default;
+
+  int i;
+  for (i = 0; i < EnabledActiveParticlesCount; i++)
+  {
+    ActiveParticleType_info *ActiveParticleTypeToEvaluate = EnabledActiveParticles[i];
+    ActiveParticleTypeToEvaluate->DescribeSupplementalData(flags);
+  }
+
+  struct ActiveParticleFormationData supplemental_data = data_default;
+  supplemental_data.level = level;
+  supplemental_data.GridID = this->ID;
+
+  ActiveParticleType::ConstructData(this, flags, supplemental_data);
+
+  /******************** FORMATION ********************/
+
+  NumberOfNewParticles = 0;
+  /* Now we iterate */
+  for (i = 0; i < EnabledActiveParticlesCount; i++)
+  {
+    ActiveParticleType_info *ActiveParticleTypeToEvaluate = EnabledActiveParticles[i];
+    ActiveParticleTypeToEvaluate->EvaluateFormation(
+                                this, supplemental_data);
+    NumberOfNewParticles += supplemental_data.NumberOfNewParticles;
+    
+  }
+
+  /* Now we copy the particles from NewParticles into a statically allocated
+   * array */
+
+  if (NumberOfNewParticles > 0) {
+    this->AddActiveParticles(supplemental_data.NewParticles, 0, NumberOfNewParticles);
+    if (debug2)
+      printf("Creating %d new active particles\n", NumberOfNewParticles);
+  }
+
+  /******************** FEEDBACK ********************/
+
+  /* Now we iterate */
+  if (NumberOfActiveParticles > 0)
+    for (i = 0; i < EnabledActiveParticlesCount; i++)
+      {
+	ActiveParticleType_info *ActiveParticleTypeToEvaluate = EnabledActiveParticles[i];
+	ActiveParticleTypeToEvaluate->EvaluateFeedback(this, supplemental_data);
+      }
+  
+  ActiveParticleType::DestroyData(this, supplemental_data);
+
+  //if (debug) printf("StarParticle: end\n");
+
+  LCAPERF_STOP("grid_ActiveParticleHandler");
+  return SUCCESS;
+}
+
+
+int grid::ActiveParticleHandler_Convert(HierarchyEntry* SubgridPointer, int level,
+                                int gridnum, int &NumberOfNewParticles)
+{
```

Comment:
Does this Convert function get used anywhere?

Reply:
Hmmm good question. The idea behind this function is that it allows for the conversion of Star Objects into Active Particles. It should be done on the command line similar to the way enzo.exe can create new outputs with the GravitationalPotential for example. I think in this case the usage was ./enzo.exe -a <snapshot> and then the new snapshop would have all of the star objects stored as active particles. It provides backwards compatibility with star objects. However, it looks like the front end part of the code is missing. I may have done that on purpose with the idea of making the frontend a standalone PR. I wrote this bit of code and it worked fine.

Reply:
This function is used in `ConvertParticles2ActiveParticles.C` line 189